File: packages/php/src/AxeneException.php
<?php

declare(strict_types=1);

namespace Axene\Mailer;

use Exception;
use ReflectionProperty;
use Throwable;

/**
 * Thrown for any non-2xx API response, or for a transport failure that
 * survives all retries.
 *
 * Inspect {@see AxeneException::getStatus()} and {@see AxeneException::getCode()}
 * to branch on specific failures (for example a 422 with code "invalid").
 *
 * Note: the inherited {@see Exception::getCode()} is final, so the API's
 * string error code is stored into the inherited `code` slot at construction.
 * That keeps the documented `getCode(): ?string` accessor working without
 * shadowing the final parent method.
 */
final class AxeneException extends Exception
{
    /** @param mixed $detail The raw parsed response body, for debugging. */
    public function __construct(
        private readonly int $status,
        string $message,
        ?string $code = null,
        private readonly mixed $detail = null,
        ?Throwable $previous = null,
    ) {
        parent::__construct($message, 0, $previous);

        // Exception::$code is loosely typed; place the string API code there so
        // the inherited final getCode() returns it.
        if ($code !== null) {
            $prop = new ReflectionProperty(Exception::class, 'code');
            $prop->setValue($this, $code);
        }
    }

    /**
     * HTTP status code. 0 indicates a transport/network failure (no response).
     */
    public function getStatus(): int
    {
        return $this->status;
    }

    /**
     * The raw parsed response body, for debugging.
     */
    public function getDetail(): mixed
    {
        return $this->detail;
    }
}

File: packages/dotnet/src/AxeneException.cs
using System;
using System.Text.Json;

namespace Axene.Mailer
{
    /// <summary>Thrown for any non-2xx API response, or a transport failure that survives all retries.</summary>
    public sealed class AxeneException : Exception
    {
        /// <summary>HTTP status code. <c>0</c> indicates a transport/network failure (no response).</summary>
        public int Status { get; }

        /// <summary>Machine-readable error code from the API body, when present.</summary>
        public string? Code { get; }

        /// <summary>Create an exception.</summary>
        public AxeneException(int status, string message, string? code) : base(message)
        {
            Status = status;
            Code = code;
        }

        /// <summary>Parse the API's <c>{ "detail": { "code", "message" } }</c> (or string) body into an exception.</summary>
        internal static AxeneException FromResponse(int status, string body)
        {
            string message = $"Axene request failed ({status})";
            string? code = null;
            try
            {
                using var doc = JsonDocument.Parse(body);
                if (doc.RootElement.TryGetProperty("detail", out var detail))
                {
                    if (detail.ValueKind == JsonValueKind.String)
                    {
                        message = detail.GetString() ?? message;
                    }
                    else if (detail.ValueKind == JsonValueKind.Object)
                    {
                        if (detail.TryGetProperty("message", out var m)) message = m.GetString() ?? message;
                        if (detail.TryGetProperty("code", out var c)) code = c.GetString();
                    }
                }
            }
            catch
            {
                // non-JSON body: keep the generic message
            }
            return new AxeneException(status, message, code);
        }
    }
}
